### ISSUE TITLE
Autodiff has problems with branches

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#3802]
**Describe the bug**
Autodiff has some problems with simple branches that depend on values of a field.

**To Reproduce**
```python
@ti.func
def frac(x):
    fractional = x - ti.floor(x)  # working but not consistent with torch.frac()
    # fractional = x - ti.floor(x) if x > 0. else x - ti.ceil(x)  # Not working, #1
    # the below is not working, either, #2
    # if x > 0.:
    #     fractional = x - ti.floor(x)
    # else:
    #     fractional = x - ti.ceil(x)
    print("x = {}, frac = {}".format(x, fractional))
    return fractional


@ti.kernel
def ti_frac(input_field: ti.template(), output_field: ti.template()):
    for i in input_field:
        output_field[i] = frac(input_field[i]) ** 2


@ti.kernel
def calc_loss(input_field: ti.template(), loss: ti.template()):
    for i in input_field:
        loss[None] += input_field[i]


def test_ti_frac_grad():
    ti.init(ti.cpu)
    torch.random.manual_seed(1)
    randoms = torch.randn(10)
    field0 = ti.field(dtype=ti.f32, shape=(10,), needs_grad=True)
    field0.from_torch(randoms)
    field1 = ti.field(dtype=ti.f32, shape=(10,), needs_grad=True)
    loss = ti.field(dtype=ti.f32, shape=(), needs_grad=True)
    with ti.Tape(loss):
        ti_frac(field0, field1)
        calc_loss(field1, loss)

    print(field0)
    print(field0.grad)
    print(torch.frac(randoms))
```

**Log/Screenshots**
SInce the log is too long and this issue can be easily reproduced, I don't paste the log here. Notably, the errors are quite different with code snippet `#1` and `#2`. Moreover, the error log with code snippet `#2` may be also related to type inference.

**Additional comments**
My `taichi` is the latest `0.8.7`.
Besides this, it will be great if you can add a `ti.frac` which is consistent with `torch.frac`.

**Comment by ailzhang** [comment]
Hey @ifsheldon, thanks a lot for reporting! I can confirm it's reproducible locally and I'll try to take a look and get back to you!

**Comment by ailzhang** [comment]
cc: @lin-hitonami for the second case which is a frontend error :D

**Comment by lin-hitonami** [comment]
Hi, @ifsheldon . Although taichi uses python as the frontend, it is lexically-scoped. That is, you cannot use a variable outside the block it is defined. So, `fractional` is not accessible outside `if` if you define it inside `if`. It will work if you write like this.
```python
    fractional = x
    if x > 0.:
        fractional -= ti.floor(x)
    else:
        fractional -= ti.ceil(x)
```

**Comment by ifsheldon** [comment]
Well then, please (update the documentation and improve a bit the error message) or implement some simple code transformations so that assignments in simple exhaustive branches, like this example, can be transformed into a nested if-else assignment.

**Comment by MarisaKirisame** [comment]
@lin-hitonami I am more confused over the explanation then before.
in particular, if taichi do not support case-2, it should be an error in a earlier pass of the compiler, (e.g. the typechecking pass), before reaching autodiff.
this also do not explain the issue with case-1

**Comment by lin-hitonami** [comment]
@MarisaKirisame The error in #2 is raised when converting the python AST to C++ AST, which is the first pass of the compiler. The error message is a bit confusing, and it indeed needs to be improved.

**Comment by erizmr** [comment]
Case 1 is fixed by #3844. Error message for case 2 has been updated.
